### DISCUSSION TITLE
[REQUEST] Use more Rule Trigger Comparison Operators for strings (contains and/or starts and ends with)

### DISCUSSION THREAD

**Discussion by ortegafernando** [arendst/Tasmota#10208] [Ideas]
### PROBLEM DESCRIPTION
_A clear and concise description of what the problem is._
HI, I am making some rules with SerialRecieved/TuyaRecieved/SSerialRecieved. It will be great to use some comparators for "ends with", "starts with" and "contains". 

But we need to use, not to have problems in "parseCompareExpression" function, for example these: "$>  $<   $|"
(as in HTML tag's for example: < is "start" and > is "end")
I understand that we will have to add more comparators and a few lines of code:

```
const char kCompareOperators[] PROGMEM = "=\0>\0<\0|\0==!=>=<=";  //  =    >    <    |    ==    !=   >=   <=
const char kCompareOperators[] PROGMEM = "=\0>\0<\0|\0==!=>=<=$>$<$|";  //  =    >    <    |    ==    !=   >=   <=   $>  $<  $|

#define COMPARE_OPERATOR_NONE            -1
#define COMPARE_OPERATOR_EQUAL            0
#define COMPARE_OPERATOR_BIGGER           1
#define COMPARE_OPERATOR_SMALLER          2
#define COMPARE_OPERATOR_EXACT_DIVISION   3
#define COMPARE_OPERATOR_NUMBER_EQUAL     4
#define COMPARE_OPERATOR_NOT_EQUAL        5
#define COMPARE_OPERATOR_BIGGER_EQUAL     6
#define COMPARE_OPERATOR_SMALLER_EQUAL    7
#define COMPARE_OPERATOR_STRING_ENDS_WITH    8    //$>
#define COMPARE_OPERATOR_STRING_STARTS_WITH    9      //$<
#define COMPARE_OPERATOR_STRING_CONTAINS      10    //$|
#define MAXIMUM_COMPARE_OPERATOR          COMPARE_OPERATOR_SMALLER_EQUAL
```
That is because I get sometimes this 55aa0005000501010001010d or this 55aa000200000155aa0005000501010001010d 

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [x ] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [x] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [x] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [x] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [x] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [] Device used (e.g., Sonoff Basic): tuya device
- [x] Tasmota binary firmware version number used: tasmota.bin (9.1.0)
  - [x] Pre-compiled
  - [ ] Self-compiled
- [ ] Flashing tools used: _____
- [ ] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:


```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:
For example: 
rule2 ON TuyaReceived#Data$>1010d DO POWER ON ENDON

```
- [ ] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:
It is a request

```
- [ ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:
It is a request

```

### TO REPRODUCE
_Steps to reproduce the behavior:_
It is a request

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._
Get managed to compare strings: ends with, starts with and contains.

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._
It is a request

### ADDITIONAL CONTEXT
_Add any other context about the problem here._
Thanks a lot for your effort

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by vic42** [comment]
I just created a driver that would benefit from **boolean** operators like bitwise `& | ^` on hex numbers.
That would possibly serve as a solution for the original requestor's problem as well.

**Reply by barbudor** [reply]
How would you use that ?
`|` is already used by the modulo operators where `value|mod` triggers the rule if `value % val == 0`
I could imagine `value&mask` would trigger the rule if `value & mask != 0`
How would you use an OR and XOR operators in rules trigger ?

**Comment by ortegafernando** [comment]
I think it is easy to solve in https://github.com/arendst/Tasmota/blob/development/tasmota/xdrv_10_rules.ino  after line 581 code similar to:

```
case COMPARE_OPERATOR_STRING_ENDS_WITH    8    //$>
     match = str_value.endsWith(rule_svalue)
     break;
case COMPARE_OPERATOR_STRING_STARTS_WITH    9      //$<
     match = str_value.startsWith(rule_svalue)
     break;
case COMPARE_OPERATOR_STRING_CONTAINS      10    //$|
     match = (str_value.indexOf(rule_svalue) > 0)
     break;
```

And in line 99 & 100:
```
#define COMPARE_OPERATOR_STRING_ENDS_WITH    8    //$>
#define COMPARE_OPERATOR_STRING_STARTS_WITH    9      //$<
#define COMPARE_OPERATOR_STRING_CONTAINS      10    //$|
#define MAXIMUM_COMPARE_OPERATOR          COMPARE_OPERATOR_STRING_CONTAINS      
const char kCompareOperators[] PROGMEM = "=\0>\0<\0|\0==!=>=<=$>$<$|";
```

Also we must to update the wiki docs.

I am not an expert in tasmota code, but I think that it will work, and probably not add to much size to binaries as functions starts_with, ends_with, ... are already used in main code.

Probably my code will need be updated so it is not case sensitive. But may be Tasmota administrators prefer it "case sensitive".

I hope Tasmota will add this feature.

Thanks a lot for your work.

**Comment by ortegafernando** [comment]
I have tried to create a PR (I am not an expert):

https://github.com/arendst/Tasmota/pull/10538

ALso, I have updated docs with a PR: https://github.com/tasmota/docs/pull/593

**Comment by barbudor** [comment]
Wouldn't a support for regex more generic ?
There is Arduino library for it (never tested)
A syntax such as `=/******/` with ***** being the regex

As per regex, 
Start with becomes `=/^start/`
End with becomes `=/end$/`
Contains becomes `=/.*content.*/`

**Reply by sfromis** [reply]
Did you check how many K that regex library would add to Tasmota? Probably quite a growth...

**Reply by ortegafernando** [reply]
I think so. Regex will add lots of KBs to Tasmota. 
What I have propossed not. I think it is a few bytes.
Thanks a lot.

**Reply by barbudor** [reply]
👍

**Comment by vic42** [comment]
One thing that is still missing is the comparison if a string is **not** equal or does **not** contain something.
The `!=` operator compares numbers only.

**Reply by barbudor** [reply]
Done #10779
